File: labodc-backend/src/main/java/com/uth/labodc/service/EnterpriseService.java
package com.uth.labodc.service;

import com.uth.labodc.dto.EnterpriseDTO;
import com.uth.labodc.model.entity.Enterprise;
import com.uth.labodc.model.entity.EnterpriseRejection;
import com.uth.labodc.model.entity.User;
import com.uth.labodc.model.enums.EnterpriseStatus;
import com.uth.labodc.model.enums.ProjectStatus;
import com.uth.labodc.repository.EnterpriseRepository;
import com.uth.labodc.repository.EnterpriseRejectionRepository;
import com.uth.labodc.repository.ProjectRepository;
import com.uth.labodc.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Service
@Slf4j
@RequiredArgsConstructor
public class EnterpriseService {
    
    private final EnterpriseRepository enterpriseRepository;
    private final EnterpriseRejectionRepository enterpriseRejectionRepository;
    private final ProjectRepository projectRepository;
    private final UserRepository userRepository;
    
    public User findUserByEmail(String email) {
        return userRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("User not found with email: " + email));
    }
    
    public List<User> searchUsers(String query) {
        return userRepository.findTop10ByEmailContainingIgnoreCaseAndDeletedAtIsNullOrderByEmailAsc(query);
    }
    
    @Transactional(readOnly = true)
    public Enterprise findById(Long id) {
        Enterprise enterprise = enterpriseRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Enterprise not found with id: " + id));
        
        // Load rejection info if enterprise was rejected
        if (EnterpriseStatus.REJECTED.equals(enterprise.getStatus())) {
            List<EnterpriseRejection> rejections = enterpriseRejectionRepository
                    .findByEnterpriseIdOrderByRejectedAtDesc(id);
            if (!rejections.isEmpty()) {
                EnterpriseRejection latestRejection = rejections.get(0);
                enterprise.setRejectionReason(latestRejection.getRejectionReason());
                enterprise.setRejectedAt(latestRejection.getRejectedAt());
                enterprise.setRejectedBy(latestRejection.getRejectedBy());
            }
        }
        
        return enterprise;
    }
    
    @Transactional(readOnly = true)
    public List<Enterprise> findAll() {
        return enterpriseRepository.findAll();
    }
    
    @Transactional
    public Enterprise verifyEnterprise(Long id, Long verifiedBy) {
        log.info("Verifying enterprise with id: {}", id);
        
        Enterprise enterprise = findById(id);
        
        if (EnterpriseStatus.APPROVED.equals(enterprise.getStatus())) {
            throw new RuntimeException("Enterprise is already approved");
        }
        
        enterprise.setStatus(EnterpriseStatus.APPROVED);
        enterprise.setVerifiedAt(LocalDateTime.now());
        enterprise.setVerifiedBy(verifiedBy);
        
        Enterprise saved = enterpriseRepository.save(enterprise);
        log.info("Enterprise {} approved successfully", id);
        
        return saved;
    }
    
    @Transactional
    public void deleteEnterprise(Long id, Long rejectedBy, String reason) {
        log.info("Rejecting enterprise with id: {}", id);
        
        Enterprise enterprise = findById(id);
        
        // Update enterprise: set status = REJECTED
        enterprise.setStatus(EnterpriseStatus.REJECTED);
        enterprise.setVerifiedAt(LocalDateTime.now());
        enterprise.setVerifiedBy(rejectedBy);
        enterpriseRepository.save(enterprise);
        log.info("Enterprise {} marked as REJECTED", id);
        
        // Save rejection record with rejection reason
        EnterpriseRejection rejection = EnterpriseRejection.builder()
                .enterpriseId(id)
                .rejectedBy(rejectedBy)
                .rejectionReason(reason)
                .rejectedAt(LocalDateTime.now())
                .companyName(enterprise.getCompanyName())
                .taxCode(enterprise.getTaxCode())
                .contactEmail(enterprise.getContactEmail())
                .build();
        
        enterpriseRejectionRepository.save(rejection);
        log.info("Rejection record saved for enterprise {} ({}): {}", id, enterprise.getCompanyName(), reason);
        
        log.info("Enterprise {} rejected successfully", id);
    }
    
    // ===== LAB ADMIN METHODS =====
    
    @Transactional(readOnly = true)
    public Page<EnterpriseDTO> getAllEnterprises(Pageable pageable) {
        return enterpriseRepository.findAll(pageable)
                .map(this::convertToDTO);
    }
    
    @Transactional(readOnly = true)
    public EnterpriseDTO getEnterpriseById(Long id) {
        Enterprise enterprise = findById(id);
        return convertToDTO(enterprise);
    }
    
    @Transactional
    public EnterpriseDTO createEnterprise(EnterpriseDTO dto, Long userId) {
        log.info("Creating new enterprise by user: {}", userId);
        
        // Check if tax code already exists
        if (dto.getTaxCode() != null && enterpriseRepository.existsByTaxCode(dto.getTaxCode())) {
            throw new RuntimeException("Mã số thuế '" + dto.getTaxCode() + "' đã tồn tại trong hệ thống");
        }
        
        // Check if user already has an enterprise
        if (enterpriseRepository.existsByUserId(userId)) {
            throw new RuntimeException("User này đã có doanh nghiệp trong hệ thống");
        }
        
        Enterprise enterprise = new Enterprise();
        enterprise.setUserId(userId);
        enterprise.setCompanyName(dto.getCompanyName());
        enterprise.setTaxCode(dto.getTaxCode());
        enterprise.setBusinessLicenseNumber(dto.getBusinessLicenseNumber());
        enterprise.setAddress(dto.getAddress());
        enterprise.setCity(dto.getCity());
        enterprise.setDistrict(dto.getDistrict());
        enterprise.setWard(dto.getWard());
        enterprise.setRepresentativeName(dto.getRepresentativeName());
        enterprise.setRepresentativePosition(dto.getRepresentativePosition());
        enterprise.setContactEmail(dto.getContactEmail());
        enterprise.setContactPhone(dto.getContactPhone());
        enterprise.setWebsite(dto.getWebsite());
        enterprise.setIndustry(dto.getIndustry());
        enterprise.setCompanySize(dto.getCompanySize());
        enterprise.setYearEstablished(dto.getYearEstablished());
        enterprise.setDescription(dto.getDescription());
        enterprise.setStatus(EnterpriseStatus.PENDING);
        enterprise.setCreatedAt(LocalDateTime.now());
        enterprise.setUpdatedAt(LocalDateTime.now());
        
        Enterprise saved = enterpriseRepository.save(enterprise);
        log.info("Enterprise created with id: {}", saved.getId());
        
        return convertToDTO(saved);
    }
    
    @Transactional
    public EnterpriseDTO updateEnterprise(Long id, EnterpriseDTO dto) {
        log.info("Updating enterprise with id: {}", id);
        
        Enterprise enterprise = findById(id);
        
        // Update fields
        enterprise.setCompanyName(dto.getCompanyName());
        enterprise.setTaxCode(dto.getTaxCode());
        enterprise.setBusinessLicenseNumber(dto.getBusinessLicenseNumber());
        enterprise.setAddress(dto.getAddress());
        enterprise.setCity(dto.getCity());
        enterprise.setDistrict(dto.getDistrict());
        enterprise.setWard(dto.getWard());
        enterprise.setRepresentativeName(dto.getRepresentativeName());
        enterprise.setRepresentativePosition(dto.getRepresentativePosition());
        enterprise.setContactEmail(dto.getContactEmail());
        enterprise.setContactPhone(dto.getContactPhone());
        enterprise.setWebsite(dto.getWebsite());
        enterprise.setIndustry(dto.getIndustry());
        enterprise.setCompanySize(dto.getCompanySize());
        enterprise.setYearEstablished(dto.getYearEstablished());
        enterprise.setDescription(dto.getDescription());
        
        // Update status if provided (Lab Admin can change status manually)
        if (dto.getStatus() != null) {
            enterprise.setStatus(EnterpriseStatus.valueOf(dto.getStatus()));
        }
        
        enterprise.setUpdatedAt(LocalDateTime.now());
        
        Enterprise saved = enterpriseRepository.save(enterprise);
        log.info("Enterprise {} updated successfully", id);
        
        return convertToDTO(saved);
    }
    
    private EnterpriseDTO convertToDTO(Enterprise enterprise) {
        // Get user email
        String userEmail = userRepository.findById(enterprise.getUserId())
                .map(User::getEmail)
                .orElse(null);
        
        // Get verifier name if verified
        String verifiedByName = null;
        if (enterprise.getVerifiedBy() != null) {
            verifiedByName = userRepository.findById(enterprise.getVerifiedBy())
                    .map(User::getEmail)
                    .orElse(null);
        }
        
        // Get project stats
        Integer totalProjects = projectRepository.countByEnterpriseId(enterprise.getId());
        Integer activeProjects = projectRepository.countByEnterpriseIdAndStatus(
                enterprise.getId(), ProjectStatus.IN_PROGRESS);
        
        return EnterpriseDTO.builder()
                .id(enterprise.getId())
                .userId(enterprise.getUserId())
                .userEmail(userEmail)
                .companyName(enterprise.getCompanyName())
                .taxCode(enterprise.getTaxCode())
                .businessLicenseNumber(enterprise.getBusinessLicenseNumber())
                .address(enterprise.getAddress())
                .city(enterprise.getCity())
                .district(enterprise.getDistrict())
                .ward(enterprise.getWard())
                .representativeName(enterprise.getRepresentativeName())
                .representativePosition(enterprise.getRepresentativePosition())
                .contactEmail(enterprise.getContactEmail())
                .contactPhone(enterprise.getContactPhone())
                .website(enterprise.getWebsite())
                .industry(enterprise.getIndustry())
                .companySize(enterprise.getCompanySize())
                .yearEstablished(enterprise.getYearEstablished())
                .description(enterprise.getDescription())
                .status(enterprise.getStatus().name())
                .verifiedAt(enterprise.getVerifiedAt())
                .verifiedBy(enterprise.getVerifiedBy())
                .verifiedByName(verifiedByName)
                .createdAt(enterprise.getCreatedAt())
                .updatedAt(enterprise.getUpdatedAt())
                .totalProjects(totalProjects)
                .activeProjects(activeProjects)
                .build();
    }
}

File: labodc_mobile/lib/services/admin/enterprise_service.dart
import 'package:dio/dio.dart';
import 'package:labodc_mobile/services/api_service.dart';
import 'package:logger/logger.dart';

/// Enterprise Management Service for Lab Admin
class EnterpriseAdminService {
  final ApiService _api = ApiService();
  final _logger = Logger();

  /// Get enterprise statistics
  Future<EnterpriseStats> getStats() async {
    try {
      _logger.i('📊 Loading enterprise stats...');
      final response = await _api.get('/enterprises/stats');
      
      final data = response.data['data'];
      _logger.i('✅ Enterprise stats loaded');
      
      return EnterpriseStats.fromJson(data);
    } catch (e) {
      _logger.e('❌ Error loading enterprise stats', error: e);
      rethrow;
    }
  }

  /// Get list of enterprises
  Future<Map<String, dynamic>> getEnterprises(Map<String, dynamic> params) async {
    try {
      _logger.i('🏢 Loading enterprises with params: $params');
      final response = await _api.get('/enterprises/management', queryParameters: params);
      
      // Backend trả về data là array trực tiếp, không có enterprises và pagination wrapper
      final data = response.data['data'];
      if (data is! List) {
        throw Exception('Invalid response format: data is not a List');
      }
      
      final enterprises = (data as List)
          .map((json) => EnterpriseListItem.fromJson(json))
          .toList();
      
      _logger.i('✅ Loaded ${enterprises.length} enterprises');
      
      return {
        'enterprises': enterprises,
        'pagination': null, // Backend chưa implement pagination
      };
    } catch (e) {
      _logger.e('❌ Error loading enterprises', error: e);
      rethrow;
    }
  }

  /// Get enterprise by ID
  Future<EnterpriseDetail> getEnterpriseById(int id) async {
    try {
      _logger.i('🏢 Loading enterprise detail: $id');
      final response = await _api.get('/enterprises/$id');
      
      final data = response.data['data'];
      _logger.i('✅ Enterprise detail loaded');
      
      return EnterpriseDetail.fromJson(data);
    } catch (e) {
      _logger.e('❌ Error loading enterprise detail', error: e);
      rethrow;
    }
  }

  /// Verify/Approve enterprise
  Future<void> verifyEnterprise(int id) async {
    try {
      _logger.i('✅ Verifying enterprise: $id');
      await _api.put('/enterprises/$id/verify');
      _logger.i('✅ Enterprise verified successfully');
    } catch (e) {
      _logger.e('❌ Error verifying enterprise', error: e);
      rethrow;
    }
  }

  /// Reject enterprise
  Future<void> rejectEnterprise(int id, String? reason) async {
    try {
      _logger.i('❌ Rejecting enterprise: $id');
      await _api.delete('/enterprises/$id', data: {'reason': reason});
      _logger.i('✅ Enterprise rejected successfully');
    } catch (e) {
      _logger.e('❌ Error rejecting enterprise', error: e);
      rethrow;
    }
  }
}

// ============= MODELS =============

class EnterpriseStats {
  final int total;
  final int pending;
  final int verified;
  final int rejected;

  EnterpriseStats({
    required this.total,
    required this.pending,
    required this.verified,
    required this.rejected,
  });

  factory EnterpriseStats.fromJson(Map<String, dynamic> json) {
    return EnterpriseStats(
      total: json['total'] ?? 0,
      pending: json['pending'] ?? 0,
      verified: json['verified'] ?? 0,
      rejected: json['rejected'] ?? 0,
    );
  }
}

class EnterpriseListItem {
  final int id;
  final int userId;
  final String companyName;
  final String taxCode;
  final String contactEmail;
  final String contactPhone;
  final String industry;
  final String companySize;
  final String status;
  final DateTime? verifiedAt;
  final DateTime createdAt;
  final int totalProjects;
  final int activeProjects;
  final double totalBudget;

  EnterpriseListItem({
    required this.id,
    required this.userId,
    required this.companyName,
    required this.taxCode,
    required this.contactEmail,
    required this.contactPhone,
    required this.industry,
    required this.companySize,
    required this.status,
    this.verifiedAt,
    required this.createdAt,
    required this.totalProjects,
    required this.activeProjects,
    required this.totalBudget,
  });

  factory EnterpriseListItem.fromJson(Map<String, dynamic> json) {
    return EnterpriseListItem(
      id: json['id'] ?? 0,
      userId: json['userId'] ?? 0,
      companyName: json['companyName'] ?? '',
      taxCode: json['taxCode'] ?? '',
      contactEmail: json['contactEmail'] ?? '',
      contactPhone: json['contactPhone'] ?? '',
      industry: json['industry'] ?? '',
      companySize: json['companySize'] ?? '',
      status: json['status'] ?? 'PENDING',
      verifiedAt: json['verifiedAt'] != null ? DateTime.parse(json['verifiedAt']) : null,
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt']) : DateTime.now(),
      totalProjects: json['totalProjects'] ?? 0,
      activeProjects: json['activeProjects'] ?? 0,
      totalBudget: (json['totalBudget'] ?? 0).toDouble(),
    );
  }
}

class EnterpriseDetail {
  final int id;
  final int userId;
  final String companyName;
  final String taxCode;
  final String? businessLicenseNumber;
  final String? address;
  final String? city;
  final String? district;
  final String? ward;
  final String? representativeName;
  final String? representativePosition;
  final String contactEmail;
  final String contactPhone;
  final String? website;
  final String industry;
  final String companySize;
  final int? yearEstablished;
  final String? description;
  final String status;
  final DateTime? verifiedAt;
  final int? verifiedBy;
  final DateTime createdAt;
  final DateTime updatedAt;
  final String? rejectionReason;
  final DateTime? rejectedAt;
  final int? rejectedBy;

  EnterpriseDetail({
    required this.id,
    required this.userId,
    required this.companyName,
    required this.taxCode,
    this.businessLicenseNumber,
    this.address,
    this.city,
    this.district,
    this.ward,
    this.representativeName,
    this.representativePosition,
    required this.contactEmail,
    required this.contactPhone,
    this.website,
    required this.industry,
    required this.companySize,
    this.yearEstablished,
    this.description,
    required this.status,
    this.verifiedAt,
    this.verifiedBy,
    required this.createdAt,
    required this.updatedAt,
    this.rejectionReason,
    this.rejectedAt,
    this.rejectedBy,
  });

  factory EnterpriseDetail.fromJson(Map<String, dynamic> json) {
    return EnterpriseDetail(
      id: json['id'] ?? 0,
      userId: json['userId'] ?? 0,
      companyName: json['companyName'] ?? '',
      taxCode: json['taxCode'] ?? '',
      businessLicenseNumber: json['businessLicenseNumber'],
      address: json['address'],
      city: json['city'],
      district: json['district'],
      ward: json['ward'],
      representativeName: json['representativeName'],
      representativePosition: json['representativePosition'],
      contactEmail: json['contactEmail'] ?? '',
      contactPhone: json['contactPhone'] ?? '',
      website: json['website'],
      industry: json['industry'] ?? '',
      companySize: json['companySize'] ?? '',
      yearEstablished: json['yearEstablished'],
      description: json['description'],
      status: json['status'] ?? 'PENDING',
      verifiedAt: json['verifiedAt'] != null ? DateTime.parse(json['verifiedAt']) : null,
      verifiedBy: json['verifiedBy'],
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt']) : DateTime.now(),
      updatedAt: json['updatedAt'] != null ? DateTime.parse(json['updatedAt']) : DateTime.now(),
      rejectionReason: json['rejectionReason'],
      rejectedAt: json['rejectedAt'] != null ? DateTime.parse(json['rejectedAt']) : null,
      rejectedBy: json['rejectedBy'],
    );
  }
}

class EnterpriseDocument {
  final int id;
  final String name;
  final String type;
  final String url;
  final DateTime uploadedAt;

  EnterpriseDocument({
    required this.id,
    required this.name,
    required this.type,
    required this.url,
    required this.uploadedAt,
  });

  factory EnterpriseDocument.fromJson(Map<String, dynamic> json) {
    return EnterpriseDocument(
      id: json['id'] ?? 0,
      name: json['name'] ?? '',
      type: json['type'] ?? '',
      url: json['url'] ?? '',
      uploadedAt: DateTime.parse(json['uploadedAt']),
    );
  }
}
